### ISSUE TITLE
Schema Registry multiple events per topic

### ISSUE THREAD

**Issue by bbejeck** [confluentinc/kafka-tutorials#624]
Schema Registry 5.5.0 supports having multiple event types per topic.  This tutorial would cover what's needed to enable this functionality and when to use it.  This feature builds off schema references,  so #623 should be done first.

**Comment by ybyzek** [comment]
If we wanted to tie this in with streaming processing, then for ksqlDB:

ksqlDB does not have much support for topics with multiple schema types right now: https://github.com/confluentinc/ksql/issues/1267.

Potential workaround from @mikebin 

That said, I think if you have a union/oneOf schema defined as in the blog post, ksqlDB will infer a “superset” schema containing all the types in the union. For example, with a schema like this:

```
{
  "fields": [
    {
      "name": "oneof_type",
      "type": [
        {
          "fields": [
            {
              "name": "fname",
              "type": "string"
            },
            {
              "name": "lname",
              "type": "string"
            }
          ],
          "name": "Customer",
          "type": "record"
        },
        {
          "fields": [
            {
              "name": "city",
              "type": "string"
            },
            {
              "name": "state",
              "type": "string"
            }
          ],
          "name": "Address",
          "type": "record"
        }
      ]
    }
  ],
  "name": "AllTypes",
  "namespace": "io.confluent.examples.avro",
  "type": "record"
}
```

ksqlDB would infer a schema that looks like this:


```
 Field      | Type
-----------------------------------------------------------------------------------------------------------------------------
 ONEOF_TYPE | STRUCT<CUSTOMER STRUCT<FNAME VARCHAR(STRING), LNAME VARCHAR(STRING)>, ADDRESS STRUCT<CITY VARCHAR(STRING), STATE VARCHAR(STRING)>>
```

So you could possibly write queries with CASE expressions to check which of the types are populated (i.e., non-null) for each record. I haven’t gone through an actual implementation of this, so perhaps someone else on the channel will have better insight and suggestions.

**Comment by ybyzek** [comment]
Related blog post: https://www.confluent.io/blog/multiple-event-types-in-the-same-kafka-topic/

How to configure the consumer: 

- https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-avro.html#avro-deserializer
- https://github.com/openweb-nl/kafka-graphql-examples/blob/307bbad6f10e4aaa6b797a3bbe3b6620d3635263/graphql-endpoint/src/main/java/nl/openweb/graphql_endpoint/service/AccountCreationService.java#L47
- https://github.com/openweb-nl/kafka-graphql-examples/blob/307bbad6f10e4aaa6b797a3bbe3b6620d3635263/graphql-endpoint/src/main/resources/application.yml#L20
- You need the generated Avro classes on the classpath. Most likely by adding a dependency.

**Comment by pbarsotti-glovo** [comment]
> Related blog post: https://www.confluent.io/blog/multiple-event-types-in-the-same-kafka-topic/
> 
> How to configure the consumer:
> 
> * https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-avro.html#avro-deserializer
> * https://github.com/openweb-nl/kafka-graphql-examples/blob/307bbad6f10e4aaa6b797a3bbe3b6620d3635263/graphql-endpoint/src/main/java/nl/openweb/graphql_endpoint/service/AccountCreationService.java#L47
> * https://github.com/openweb-nl/kafka-graphql-examples/blob/307bbad6f10e4aaa6b797a3bbe3b6620d3635263/graphql-endpoint/src/main/resources/application.yml#L20
> * You need the generated Avro classes on the classpath. Most likely by adding a dependency.

I think all the links that you shared on how to configure the consumer are related to `TopicRecordNameStrategy`, as you can see [here](https://github.com/openweb-nl/kafka-graphql-examples/blob/307bbad6f10e4aaa6b797a3bbe3b6620d3635263/graphql-endpoint/src/main/resources/application.yml#L29). To use the approach described in Yokota's post, `TopicNameStrategy` must be used, with the wrapper event schema defined in the topic (the one containing the UNION).

On the other hand, can you confirm if this alternative works for ksqlDB? I'm in an early stage deciding if I go with this approach or the `TopicRecordNameStrategy` and the only thing making me doubt about using the later is ksqlDB.
This mechanism using references does not allow specific records usage so I have to do a lot of inferences and transformations to map a GenericRecord to one of the possible SpecificRecords contained inside.
Schema evolution is also harder because you need to change the child schema and update the version on the parent.
I don't want to invest a lot of time on it if ksqlDB does not work well with this...

**Comment by ybyzek** [comment]
@bbejeck can we close this GH issue, or is there anything else to add to the KT?

**Comment by bbejeck** [comment]
@ybyzek I think so.  I'll close it and if something comes up we can always create a new issue.

**Comment by bbejeck** [comment]
Resolved via #983
